### DISCUSSION TITLE
commonjs dynamic import esm not work

### DISCUSSION THREAD

**Discussion by pincman** [swc-project/swc#8046] [General]
tsc+`"moduleResolution": "Node16"`  work fine,
but change to swc get error

Code

```typescript
async tree(@Query() options: QueryCategoryTreeDto) {
        console.log((await import('chalk')).default.bgCyan('test'));
        return this.service.findTrees(options);
    }
```
error
```
[Nest] 58863  - 2023/10/02 17:17:16   ERROR [ExceptionsHandler] require() of ES Module /Users/pincman/Code/classroom/nestapp/node_modules/.pnpm/find-up@6.3.0/node_modules/find-up/index.js from /Users/pincman/Code/classroom/nestapp/dist/modules/content/controllers/category.controller.js not supported.
Instead change the require of index.js in /Users/pincman/Code/classroom/nestapp/dist/modules/content/controllers/category.controller.js to a dynamic import() which is available in all CommonJS modules.
```

swc config
```json
{
    "$schema": "https://json.schemastore.org/swcrc",
    "sourceMaps": true,
    "module": {
        // You can specify "commonjs", "es6", "amd", "umd"
        "type": "commonjs",
        "lazy": true,
        "importInterop": "node",
        "ignoreDynamic": true
    },
    "jsc": {
        "parser": {
            "syntax": "typescript",
            "decorators": true,
            "dynamicImport": true
        },
        "transform": {
            "legacyDecorator": true,
            "decoratorMetadata": true
        },
        "baseUrl": "./"
    },
    "minify": false
}
```

**Comment by JSerFeng** [comment]
You transformed esm to cjs, which means import() becomes Promise.resolve().then(() => require()).

If you really need to transform it to cjs, set module.ignoreDynamic to true, it will preserve dynamic import expression.

**Reply by pincman** [reply]
its right,work fine now,thx!

**Reply by Xe11o** [reply]
> You transformed esm to cjs, which means import() becomes Promise.resolve().then(() => require()).
> 
> If you really need to transform it to cjs, set module.ignoreDynamic to true, it will preserve dynamic import expression.

This 'works' in that it preserves the dynamic import, but this is not compatible with the goal of transforming to cjs - you then have an 'import' statement in the middle of otherwise cjs code. Is it just not possible to do this transformation then? My problem is that with the transformation to Promise.resolve().then(() => require()), Jest errors out with a SyntaxError: Unexpected identifier 'Promise'

**Reply by carlnc** [reply]
> SyntaxError: Unexpected identifier 'Promise'

I was getting this error, until I changed lazy to true.

**Comment by sosoba** [comment]
You can use built in [styleText](https://nodejs.org/docs/latest-v22.x/api/util.html#utilstyletextformat-text-options) instead of chalk.
